include/common/robo_claw.h

```diff
@@ -0,0 +1,383 @@
+    }
+
+     //! Write dword
+    template<typename ...Data>
+    void write(CRC &crc,
+               uint32_t dword, Data... data)
+    {
+        // Write bytes of dword followed by remainder
+        // **NOTE** RoboClaw is big endian so SerialInterface::write wouldn't work
+        write(crc, (uint8_t)(dword >> 24), (uint8_t)(dword >> 16), (uint8_t)(dword >> 8),(uint8_t)dword, data...);
+    }
+
+
+    //! Write command
+    template<typename ...Data>
+    void writeCommand(Command command, Data... data)
+    {
+        for(int r = 0; r < m_MaxRetry; r++) {
+            // Start CRC calculation
+            CRC crc;
+
+            // Write data
+            write(crc, m_Address, static_cast<uint8_t>(command), data...);
+
+            // Write CRC
+            // **NOTE** RoboClaw is big endian so SerialInterface::write wouldn't work
+            m_SerialInterface.writeByte(crc.get() >> 8);
+            m_SerialInterface.writeByte(crc.get());
+
+            // If byte is read successfully
+            uint8_t response;
+            if(m_SerialInterface.readByte(response) && response == 0xFF) {
+                return;
+            }
+        }
+
+        throw std::runtime_error("Unable to write to serial port");
+    }
+
+    //! End-stop of variadic expandion
+    void read(CRC&)
+    {
+    }
+
+    //! Read a byte
+    template<typename ...Data>
+    void read(CRC &crc,
+              uint8_t *byte, Data... data)
+    {
+        // Read byte and update CRC
+        if(m_SerialInterface.readByte(*byte)) {
+            crc.update(*byte);
+        }
+        else {
+            throw std::runtime_error("Unable to read byte");
+        }
+
+        // Read remainder
+        read(crc, data...);
+    }
+
+    //! Read a word
+    template<typename ...Data>
+    void read(CRC &crc,
+              uint16_t *word, Data... data)
+    {
+        // Read two bytes followed by remainder
+        // **NOTE** RoboClaw is big endian so SerialInterface::read wouldn't work
+        uint8_t bytes[2];
+        read(crc, &bytes[0], &bytes[1], data...);
+
+        // Re-assemble word
+        *word = (bytes[0] << 8) | bytes[1];
+    }
+
+    //! Read a dword
+    template<typename ...Data>
+    void read(CRC &crc,
+              uint32_t *dword, Data... data)
+    {
+        // Read four bytes followed by remainder
+        // **NOTE** RoboClaw is big endian so SerialInterface::read wouldn't work
+        uint8_t bytes[4];
+        read(crc, &bytes[0], &bytes[1], &bytes[2], &bytes[3], data...);
+
+        // Re-assemble dword
+        *dword = (bytes[0] << 24) | (bytes[1] << 16) | (bytes[2] << 8) | bytes[3];
+    }
+
+    // Perform a read command consisting of:
+    // write address [1 byte]
+    // write command[1 byte]
+    // read variadic parameters
+    // read CRC [2 bytes]
+    template<typename ...Data>
+    void readCommand(Command command, Data... data)
+    {
+        for(int r = 0; r < m_MaxRetry; r++) {
```

Comment:
It looks like this loop is redundant: either you ``return`` or an exception is thrown.

Reply:
I suppose you could try again if the CRC check fails... Not sure if that would make sense though

Reply:
This is an excellent point! The retry logic was copied from the original arduino code but clearly doesn't currently quite make sense - with USB serial (rather than two wires jammed into a UART) I suspect it's entirely unnecessary but will check exactly when it was supposed to retry